mmv1/products/vertexai/DeploymentResourcePool.yaml

```diff
@@ -0,0 +1,175 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+--- !ruby/object:Api::Resource
+name: 'DeploymentResourcePool'
+base_url: 'projects/{{project}}/locations/{{region}}/deploymentResourcePools'
+create_url: 'projects/{{project}}/locations/{{region}}/deploymentResourcePools'
+self_link: 'projects/{{project}}/locations/{{region}}/deploymentResourcePools/{{deployment_resource_pool_id}}'
+# No method is not defined for DeploymentResourcePools
+immutable: true
+create_verb: :POST
+references: !ruby/object:Api::Resource::ReferenceLinks
+  api: https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.deploymentResourcePools
+async: !ruby/object:Api::OpAsync
+  actions:
+    - create
+    - delete
+  operation: !ruby/object:Api::OpAsync::Operation
+    path: 'name'
+    base_url: '{{op_id}}'
+    wait_ms: 1000
+  result: !ruby/object:Api::OpAsync::Result
+    path: 'response'
+    resource_inside_response: true
+  status: !ruby/object:Api::OpAsync::Status
+    path: 'done'
+    complete: true
+    allowed:
+      - true
+      - false
+  error: !ruby/object:Api::OpAsync::Error
+    path: 'error'
+    message: 'message'
+description: |-
+  'DeploymentResourcePool can be shared by multiple deployed models,
+  whose underlying specification consists of dedicated resources.'
+examples:
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'vertex_ai_deployment_resource_pool'
+    primary_resource_id: 'deployment_resource_pool'
+    # Skip the import test because deployment_resource_pool_id can't be specified in an auto-generated test
+    skip_import_test: true
+parameters:
+  - !ruby/object:Api::Type::String
+    name: region
+    description: The region of deployment resource pool. eg us-central1
+    url_param_only: true
+    immutable: true
+properties:
+  - !ruby/object:Api::Type::String
+    name: 'deploymentResourcePoolId'
+    required: true
+    # GET doesn't return id, and it returns deploymentResourcePool directly
+    ignore_read: true
+    description:
+      The ID to use for the DeploymentResourcePool, which will become the final component of the DeploymentResourcePool's resource name.
+      The maximum length is 63 characters, and valid characters are `/^[a-z]([a-z0-9-]{0,61}[a-z0-9])?$/`.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'deploymentResourcePool'
+    description: The deployment resource pool to create.
+    required: true
+    # Deal with deploymentResourcePool as a flatten object because GET returns DeploymentResourcePool
+    flatten_object: true
```

Comment:
[MEMO] The `get` returns `DeploymentResourcePool`, but `create` requires a wrapper of `DeploymentResourcePool`.

* https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.deploymentResourcePools/get
* https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.deploymentResourcePools/create

Reply:
Instead of `flatten_object` can we use an `encoder`? Very similar to something like https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/encoders/wrap_object_with_template_id.go.erb

(We likely need to remove `deploymentResourcePoolId` from `obj` and reinsert it to `obj`, I think that that encoder's `resourceId` field was `url_param_only`)

Reply:
Thank you for your feedback. I updated the code to use the custom encoder in 99bb911.